docs/proposals/petset.md

```diff
@@ -0,0 +1,533 @@
+one pod and zero or more **supporting resources** for each identity. There should never be more than one pod
+or unique supporting resource per identity at any one time. A new pod can be created for an identity only
+if a previous pod has been fully terminated (reached its graceful termination limit or cleanly exited).
+
+```
+type PetSet struct {
+    ObjectMeta
+
+    Spec PetSetSpec
+    ...
+}
+
+type PetSetSpec struct {
+    // Number of instances in the set
+    Replicas *int
+}
+```
+
+Like a replication controller, a PetSet may be targeted by an autoscaler. The PetSet makes no assumptions
+about upgrading or altering the pods in the set (similar to a DaemonSet) - instead, the user can trigger
+graceful deletion and the PetSet will replace the terminated member with the newer template once it exits.
+Future proposals may offer update capabilities. A PetSet requires RunAlways pods.
+
+### How identities are managed
+
+A key question is whether scaling down a PetSet and then scaling it back up should reuse identities. If not,
+scaling down becomes a destructive action (an admin cannot recover by scaling back up). Given the safety
+first assumption, identity reuse seems the correct default. This implies that identity assignment should
+be deterministic and not subject to controller races (a controller that has crashed during scale up should
+assign the same identities on restart, and two concurrent controllers should decide on the same outcome
+identities).
+
+The simplest way to manage identities, and easiest to understand for users, is a numeric identity system
+starting at I=1 that ranges up to the current replica count and is contiguous.
+
+### Controller behavior.
+
+When a PetSet is scaled up, the controller must create both pods and supporting resources for
+each new identity. The controller must create supporting resources for the pod before creating the
+pod. If a supporting resource with the appropriate name already exists, the controller should treat that as
+creation succeeding. If a supporting resource cannot be created, the controller should flag an error to
+status, back-off (like a scheduler or replication controller), and try again later.
+
+When a PetSet is scaled down, the pod for the removed indentity should be deleted. It is less clear what the
+controller should do to supporting resources. If every pod requires a PV, and a user accidentally scales
+up to N=200 and then back down to N=3, leaving 197 PVs lying around may be undesirable (potential for
+abuse). On the other hand, a cluster of 5 that is accidentally scaled down to 3 might irreparably destroy
+the cluster if the PV for identities 4 and 5 are deleted (may not be recoverable). For the initial proposal,
+leaving the supporting resources is the safest path (safety first) with a potential future policy applied
+to the PetSet for how to manage supporting resources (DeleteImmediately, GarbageCollect, Preserve).
```

Comment:
In the initial description, my assumption is the PVC is not released.
Releasing that asset is dangerous in many scenarios.

On Thu, Dec 10, 2015 at 10:59 AM, Matthew Wringe <notifications@github.com>
wrote:

> In docs/proposals/petset.md
> <https://github.com/kubernetes/kubernetes/pull/18016#discussion_r47243979>
> :
>
> > +
> > +### Controller behavior.
> > +
> > +When a PetSet is scaled up, the controller must create both pods and supporting resources for
> > +each new identity. The controller must create supporting resources for the pod before creating the
> > +pod. If a supporting resource with the appropriate name already exists, the controller should treat that as
> > +creation succeeding. If a supporting resource cannot be created, the controller should flag an error to
> > +status, back-off (like a scheduler or replication controller), and try again later.
> > +
> > +When a PetSet is scaled down, the pod for the removed indentity should be deleted. It is less clear what the
> > +controller should do to supporting resources. If every pod requires a PV, and a user accidentally scales
> > +up to N=200 and then back down to N=3, leaving 197 PVs lying around may be undesirable (potential for
> > +abuse). On the other hand, a cluster of 5 that is accidentally scaled down to 3 might irreparably destroy
> > +the cluster if the PV for identities 4 and 5 are deleted (may not be recoverable). For the initial proposal,
> > +leaving the supporting resources is the safest path (safety first) with a potential future policy applied
> > +to the PetSet for how to manage supporting resources (DeleteImmediately, GarbageCollect, Preserve).
>
> A supporting resource is not just the PV but the PVC as well. I would
> expect for a lot of situations that PVs would be configured externally, and
> not part of the applications deployment itself.
>
> I am trying to figure out what would happen in the situation where I have
> 3 instances, then I scale down to 2. Is the PVC released in this situation?
> or ignored since its a supporting resource?
>
> If I then scale back to 3 instances, does my new pod get the old existing
> PVC? or a new one? The counter based identifier in this case would be 3, so
> if I am trying to create a new PVC using that in my template I may run into
> issues.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/18016/files#r47243979>.
>